WikiWatershed/rapid-watershed-delineation
# Update API to support NHD

This PR is to continue work on https://github.com/WikiWatershed/rapid-watershed-delineation/pull/36

In addition to @kdeloach's work, it...

- Adds a final re-projection to LatLng to the output shapefiles
- Removes the commit that added a flask debug flag ( [see comment](https://github.com/WikiWatershed/rapid-watershed-delineation/pull/36/files#r85942343) ) 
- Fixes up the initial `rwd-nhd` endpoint commit to specify the NHD data's path into a subdirectory `/nhd`
( [see comment](https://github.com/WikiWatershed/rapid-watershed-delineation/pull/36/files#r85941746) )

### Testing

- Checkout this branch

- Checkout WikiWatershed/docker-rwd#19

If you're on a mac:
- You'll need [Docker for Mac](https://docs.docker.com/engine/installation/mac/), or the Docker toolbox. `docker-machine start` will start a VM called `default`
- I followed the steps in the second answer [here](http://stackoverflow.com/questions/30040708/how-to-mount-local-volumes-in-docker-machine) to give the the `docker-machine` vm access to the RWD data and repo
- Instead of `localhost` you'll use whatever IP `docker-machine ip default` spits out

- From the `docker-rwd` project, run:
```
./scripts/build.sh
RWD_REPO=<where ever you put the repo> RWD_DATA=<where ever you put the data> ./scripts/run.sh
```
- **The RWD NHD data folders need to be in a subdirectory called `nhd`**

Run:

`curl http://localhost:5000/rwd-nhd/39.9526/-75.1652` <= Inside the DRB
`curl http://localhost:5000/rwd/39.9526/-75.1652`  <= Inside the DRB
`curl http://localhost:5000/rwd-nhd/39.0458/-76.6413` <= Outside DRB
`curl http://localhost:5000/rwd/39.0458/-76.6413` <= Outside DRB, should raise appropriate exception


Example output
```
HTTP/1.1 200 OK
Connection: close
Content-Length: 3488
Content-Type: application/json
Date: Tue, 01 Nov 2016 16:04:59 GMT
Server: gunicorn/19.4.5

{
    "input_pt": {
        "features": [
            {
                "geometry": {
                    "coordinates": [
                        -75.16513888892709,
                        39.95337962962963
                    ],
                    "type": "Point"
                },
                "properties": {
                    "Dist_moved": 8,
                    "ID": 1,
                    "Lat": 39.9526,
                    "Lon": -75.1652
                },
                "type": "Feature"
            }
        ],
        "type": "FeatureCollection"
    },
    "watershed": {
        "crs": {
            "properties": {
                "name": "urn:ogc:def:crs:OGC:1.3:CRS84"
            },
            "type": "name"
        },
        "features": [
            {
                "geometry": {
                    "coordinates": [
                        [
                            [
                                -77.52171346605734,
                                42.66530214748923
                            ],
                            ....
                            [
                                -77.52135223128366,
                                42.665250284007094
                            ],
                            
                       ]
                    ],
                    "type": "Polygon"
                },
                "properties": {
                    "Area": 55.1538,
                    "AvgOLF": 0.012716,
                    "BasinLen": 15.537951,
                    "DrnDen": 39.320237,
                    "GRIDCODE": 1,
                    "StrLen": 2168.6605,
                    "Strord": 8.0
                },
                "type": "Feature"
            }
        ],
        "type": "FeatureCollection"
    }
}
```
